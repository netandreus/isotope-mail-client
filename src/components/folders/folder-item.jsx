import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {translate} from 'react-i18next';
import {FolderTypes} from '../../services/folder';
import styles from './folder-item.scss';
import mainCss from '../../styles/main.scss';

const ACTIONS_MENU_ANIMATION_DURATION_MS = 500;

export class FolderItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dragOver: false,
      contextMenuAnimation: false,
      contextMenuVisible: false
    };
    this.handleOnDragStart = this.onDragStart.bind(this);
    this.handleOnDragOver = this.onDragOver.bind(this);
    this.handleOnDragLeave = this.onDragLeave.bind(this);
    this.handleOnDrop = this.onDrop.bind(this);
  }

  render() {
    const {
      t, className, selected, draggable, graphic, label, newMessageCount, unreadMessageCount, onClick,
      onRename, onAddChild, onDelete
    } = this.props;
    const {contextMenuVisible, contextMenuAnimation, dragOver} = this.state;
    const labelWithCount = `${label} ${unreadMessageCount > 0 ? `(${unreadMessageCount})` : ''}`;
    const hasContextMenu = onDelete !== null || onAddChild !== null || onRename !== null;
    const conditionalDraggableAttributes = {};
    if (draggable) {
      conditionalDraggableAttributes.draggable = true;
      // Support for MS Edge DnD requires link to have a href attribute
      conditionalDraggableAttributes.href = '#';
      conditionalDraggableAttributes.onDragStart = this.handleOnDragStart;
    }
    return (
      <a className={`${className} ${mainCss['mdc-list-item']} ${mainCss['mdc-list-item__folder-item']}
        ${styles.listItem}
        ${selected ? mainCss['mdc-list-item--activated'] : ''}
        ${dragOver ? mainCss['mdc-list-item--selected'] : ''}`}
      title={labelWithCount}
      onClick={onClick}
      {...conditionalDraggableAttributes}
      onDrop={this.handleOnDrop} onDragOver={this.handleOnDragOver} onDragLeave={this.handleOnDragLeave}
      >
        <span className={`material-icons ${mainCss['mdc-list-item__graphic']} ${styles.graphic}`}>
          {graphic}
        </span>
        <span className={`${mainCss['mdc-list-item__primary-text']} ${styles.primaryText}
          ${newMessageCount > 0 ? styles.hasNewMessages : ''}`}>
          {labelWithCount}
        </span>
        <span className={`${styles.actions} ${contextMenuVisible ? styles.visible : ''} ${contextMenuAnimation ? styles.animating : ''}`}
          onMouseLeave={event => this.hideContextMenu(event)}>
          <span className={`${styles.contextMenu}`}>
            {onDelete !== null && <i
              isotip={t('sideBar.folderList.delete')} isotip-position='bottom-end' isotip-size='small'
              className={'material-icons'} onClick={onDelete}>delete</i>}
            {onAddChild !== null && <i
              isotip={t('sideBar.folderList.newSubFolder')} isotip-position='bottom-end' isotip-size='small'
              className={'material-icons'} onClick={onAddChild}>add</i>}
            {onRename !== null && <i
              isotip={t('sideBar.folderList.rename')} isotip-position='bottom-end' isotip-size='small'
              className={'material-icons'} onClick={onRename}>edit</i>}
          </span>
          {hasContextMenu &&
          <span className={styles.menuButton}>
            <i className={'material-icons'}
              isotip={t('sideBar.folderList.folderActions')} isotip-position='bottom-end' isotip-size='small'
              onClick={event => this.showContextMenu(event)}>more_vert</i>
          </span>
          }
        </span>
      </a>
    );
  }

  onDragStart(event) {
    event.stopPropagation();
    this.props.onDragStart(event);
  }

  onDrop(event) {
    event.preventDefault();
    event.stopPropagation();
    this.setState({dragOver: false});
    this.props.onDrop(event);
  }

  onDragOver(event) {
    event.preventDefault();
    event.stopPropagation();
    if (event.dataTransfer.types && Array.from(event.dataTransfer.types).includes('application/json')) {
      this.setState({dragOver: true});
    }
  }

  onDragLeave(event) {
    event.preventDefault();
    event.stopPropagation();
    this.setState({dragOver: false});
  }

  showContextMenu(event) {
    event.preventDefault();
    event.stopPropagation();
    this.setState({contextMenuVisible: true, contextMenuAnimation: true});
    setTimeout(() => this.setState({contextMenuAnimation: false}), ACTIONS_MENU_ANIMATION_DURATION_MS);
  }

  hideContextMenu() {
    this.setState({contextMenuVisible: false, contextMenuAnimation: false});
  }
}

FolderItem.propTypes = {
  className: PropTypes.string,
  graphic: PropTypes.string,
  label: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  draggable: PropTypes.bool,
  onDragStart: PropTypes.func,
  onDrop: PropTypes.func,
  onClick: PropTypes.func,
  onRename: PropTypes.func,
  onAddChild: PropTypes.func,
  onDelete: PropTypes.func,
  unreadMessageCount: PropTypes.number,
  newMessageCount: PropTypes.number
};

FolderItem.defaultProps = {
  className: '',
  graphic: FolderTypes.FOLDER.icon,
  selected: false,
  draggable: false,
  unreadMessageCount: 0,
  newMessageCount: 0,
  onDragStart: () => {},
  onDrop: null,
  onRename: null,
  onAddChild: null,
  onDelete: null
};

export default translate()(FolderItem);
