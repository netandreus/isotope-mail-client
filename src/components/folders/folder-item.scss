@import '../../styles/partials/variables';

.listItem {
  cursor: pointer;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: flex;

  .graphic {
    margin-right: $default-spacing;
  }

  .primaryText {
    margin: 0px;
    width: 1px;
    min-width: 1px;
    flex: 1;

    &.hasNewMessages {
      font-weight: $font-weight-bold;
    }
  }

  .actions {
    display: none;
    margin-right: -6px;

    .contextMenu {
      display:none;
      transform: translateX(100%);
      margin: 0;
      padding: 0;
    }

    &.animating {
      overflow: hidden;
    }

    &.visible {
      $actions-menu-animation-duration: 0.5s; // Must match with component setTimeout for animation

      .menuButton {
        display: none;
      }
      .contextMenu {
        display: block;
        margin: -16px 0 -16px -16px;
        padding: 16px 0 16px 16px;
        max-height: 100%;

        transform: translateX(0%);
        animation: slide-in $actions-menu-animation-duration forwards;
        @keyframes slide-in {
          0% { transform: translateX(100%); }
          99% { transform: translateX(0%); }
        }
      }
    }
  }

  &:hover {
    .actions {
      display: block;
    }
  }
}
