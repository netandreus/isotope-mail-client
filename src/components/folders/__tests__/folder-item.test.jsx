import React from 'react';
import {shallow} from 'enzyme/build/index';
import {FolderItem} from '../folder-item';

describe('FolderItem component test suite', () => {
  describe('Snapshot render', () => {
    test('Defaults, Should render FolderItem with no menu and no actions', () => {
      // Given
      const props = {
        t: key => key,
        className: 'the-one-percent', graphic: 'Andy Warhol', label: 'Red Label',
        selected: true, draggable: true, unreadMessageCount: 13, newMessageCount: 37};

      // When
      const folderItem = shallow(<FolderItem {...props} />);

      // Then
      expect(folderItem).toMatchSnapshot();
    });
    test('Defaults with actions, Should render FolderItem with hidden menu and actions', () => {
      // Given
      const props = {
        t: key => key,
        className: 'the-one-percent', graphic: 'Andy Warhol', label: 'Red Label',
        selected: true, draggable: true, unreadMessageCount: 13, newMessageCount: 37,
        onRename: () => {}, onDelete: () => {}};

      // When
      const folderItem = shallow(<FolderItem {...props} />);

      // Then
      expect(folderItem).toMatchSnapshot();
    });
    test('Context Menu Visible, Should render FolderItem with menu', () => {
      // Given
      const props = {
        t: key => key,
        className: 'the-one-percent', graphic: 'Andy Warhol', label: 'Red Label',
        selected: true, draggable: true, unreadMessageCount: 13, newMessageCount: 37,
        onRename: () => {}, onDelete: () => {}};

      // When
      const folderItem = shallow(<FolderItem {...props} />);
      folderItem.setState({contextMenuVisible: true});

      // Then
      expect(folderItem).toMatchSnapshot();
    });
  });
  describe('Component events', () => {
    test('Generic mouse events, events fired', () => {
      // Given
      const onClick = jest.fn();
      const onDragStart = jest.fn();
      const onDrop = jest.fn();
      const folderItem = shallow(<FolderItem label={''} selected={false} draggable={true}
        onClick={onClick} onDragStart={onDragStart} onDrop={onDrop}/>);

      // When
      folderItem.find('.listItem').simulate('click');
      folderItem.find('.listItem').simulate('dragStart', {stopPropagation: () => {}});
      folderItem.find('.listItem').simulate('drop', {preventDefault: () => {}, stopPropagation: () => {}});

      // Then
      expect(onClick).toHaveBeenCalledTimes(1);
      expect(onDragStart).toHaveBeenCalledTimes(1);
      expect(onDrop).toHaveBeenCalledTimes(1);
    });
    test('Folder dragged over, should set dragOver true', () => {
      // Given
      const props = {label: 'I\'m NOT dragged over', selected: false};
      const folderItem = shallow(<FolderItem {...props} />);

      // When
      folderItem.find('.listItem').simulate('dragOver', {
        preventDefault: () => {}, stopPropagation: () => {},
        dataTransfer: {types: ['application/hal+json', 'text/plain', 'application/json']}});

      // Then
      expect(folderItem.state('dragOver')).toEqual(true);
    });
    test('Folder drag leave, was dragged over, should set dragOver false', () => {
      // Given
      const props = {label: 'I\'m dragged over', selected: false};
      const folderItem = shallow(<FolderItem {...props} />);
      folderItem.setState({dragOver: true});

      // When
      folderItem.find('.listItem').simulate('dragLeave', {preventDefault: () => {}, stopPropagation: () => {}});

      // Then
      expect(folderItem.state('dragOver')).toEqual(false);
    });
    test('Menu clicked, should show menu', () => {
      // Given
      const props = {
        t: key => key,
        label: 'My menu is hidden', selected: true, onRename: () => {}, onDelete: () => {}};
      const folderItem = shallow(<FolderItem {...props} />);

      // When
      folderItem.find('.listItem .actions').find({children: 'more_vert'})
        .simulate('click', {preventDefault: () => {}, stopPropagation: () => {}});

      // Then
      expect(folderItem.state('contextMenuVisible')).toEqual(true);
    });
    test('Mouse Leave, menu visible, should hide menu', () => {
      // Given
      const props = {
        t: key => key,
        label: 'Menu will hide', selected: true, onRename: () => {}, onDelete: () => {}};
      const folderItem = shallow(<FolderItem {...props} />);
      folderItem.setState({contextMenuVisible: true});

      // When
      folderItem.find('.listItem .actions').simulate('mouseLeave');

      // Then
      expect(folderItem.state('contextMenuVisible')).toEqual(false);
    });
  });
});
