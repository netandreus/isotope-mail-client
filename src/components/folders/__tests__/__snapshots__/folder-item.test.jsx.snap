// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderItem component test suite Snapshot render Context Menu Visible, Should render FolderItem with menu 1`] = `
<a
  className="the-one-percent mdc-list-item mdc-list-item__folder-item
        listItem
        mdc-list-item--activated
        "
  draggable={true}
  href="#"
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  title="Red Label (13)"
>
  <span
    className="material-icons mdc-list-item__graphic graphic"
  >
    Andy Warhol
  </span>
  <span
    className="mdc-list-item__primary-text primaryText
          hasNewMessages"
  >
    Red Label (13)
  </span>
  <span
    className="actions visible "
    onMouseLeave={[Function]}
  >
    <span
      className="contextMenu"
    >
      <i
        className="material-icons"
        isotip="sideBar.folderList.delete"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        delete
      </i>
      <i
        className="material-icons"
        isotip="sideBar.folderList.rename"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        edit
      </i>
    </span>
    <span
      className="menuButton"
    >
      <i
        className="material-icons"
        isotip="sideBar.folderList.folderActions"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        more_vert
      </i>
    </span>
  </span>
</a>
`;

exports[`FolderItem component test suite Snapshot render Defaults with actions, Should render FolderItem with hidden menu and actions 1`] = `
<a
  className="the-one-percent mdc-list-item mdc-list-item__folder-item
        listItem
        mdc-list-item--activated
        "
  draggable={true}
  href="#"
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  title="Red Label (13)"
>
  <span
    className="material-icons mdc-list-item__graphic graphic"
  >
    Andy Warhol
  </span>
  <span
    className="mdc-list-item__primary-text primaryText
          hasNewMessages"
  >
    Red Label (13)
  </span>
  <span
    className="actions  "
    onMouseLeave={[Function]}
  >
    <span
      className="contextMenu"
    >
      <i
        className="material-icons"
        isotip="sideBar.folderList.delete"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        delete
      </i>
      <i
        className="material-icons"
        isotip="sideBar.folderList.rename"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        edit
      </i>
    </span>
    <span
      className="menuButton"
    >
      <i
        className="material-icons"
        isotip="sideBar.folderList.folderActions"
        isotip-position="bottom-end"
        isotip-size="small"
        onClick={[Function]}
      >
        more_vert
      </i>
    </span>
  </span>
</a>
`;

exports[`FolderItem component test suite Snapshot render Defaults, Should render FolderItem with no menu and no actions 1`] = `
<a
  className="the-one-percent mdc-list-item mdc-list-item__folder-item
        listItem
        mdc-list-item--activated
        "
  draggable={true}
  href="#"
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  title="Red Label (13)"
>
  <span
    className="material-icons mdc-list-item__graphic graphic"
  >
    Andy Warhol
  </span>
  <span
    className="mdc-list-item__primary-text primaryText
          hasNewMessages"
  >
    Red Label (13)
  </span>
  <span
    className="actions  "
    onMouseLeave={[Function]}
  >
    <span
      className="contextMenu"
    />
  </span>
</a>
`;
